### DISCUSSION TITLE
What's the pattern for a language with optional semi colons?

### DISCUSSION THREAD

**Discussion by trusktr** [eclipse-langium/langium#959] [Q&A]
What may a grammar look like for a language with optional semi colons?

F.e. suppose you want to allow the following, where all of them are a function call:

```js
foo (arg1, arg2);
// or
foo (arg1, arg2)
// or
foo
(arg1, arg2);
// or
foo
(arg1, arg2)
```

---

On a related note, I'm interested in code re-use. Are there any grammar files published somewhere that we can `import` and may have certain useful things for C-like or JS-like languages?

**Comment by msujew** [comment]
> What may a grammar look like for a language with optional semi colons?

You can take a look at the [langium-lox](https://github.com/langium/langium-lox/blob/main/src/language-server/lox.langium) grammar. Even though it contains semi colons (due to them being mandatory in the original grammar), I just confirmed that the language also works as expected when making them optional.

> On a related note, I'm interested in code re-use. Are there any grammar files published somewhere that we can import and may have certain useful things for C-like or JS-like languages?

Not yet, though we are thinking about that. We do have a few example languages, but nothing concrete such as a sort of "standard library" of Langium grammars.

**Reply by trusktr** [reply]
How does that look like? Can you perhaps push optional semis to a new branch in langium-lox, or paste the grammar here (or the diff)?

**Reply by msujew** [reply]
Sure, the changes are fairly minimal:

```diff
LoxElement: 
    Class |
    ExpressionBlock | 
    IfStatement |
    WhileStatement |
    ForStatement |
    FunctionDeclaration |
-    VariableDeclaration ';' | 
-    PrintStatement ';' | 
-    ReturnStatement ';' | 
-    Expression ';'
+    VariableDeclaration ';'? | 
+    PrintStatement ';'? | 
+    ReturnStatement ';'? | 
+    Expression ';'?
;
```

There are a few other places, such as `for` loops, which do have semi colons. However, they are mandatory for unambiguous parsing. Languages that do not use `;`, usually opt for a `for ... in ... { ... }` syntax anyway instead of the `for ;;` syntax.
